planner/core/logical_plan_builder.go

```diff
@@ -4809,6 +4854,17 @@ func (b *PlanBuilder) groupWindowFuncs(windowFuncs []*ast.WindowFuncExpr) (map[*
 			groupedWindow[updatedSpec] = append(groupedWindow[updatedSpec], windowFunc)
 		}
 	}
+	// Unused window specs should also be checked in b.buildWindowFunctions,
+	// so we add them to `groupedWindow` with empty window functions.
+	for _, spec := range b.windowSpecs {
+		if _, ok := groupedWindow[spec]; ok {
+			continue
+		}
+		if _, ok := updatedSpecMap[spec.Name.L]; ok {
+			continue
+		}
+		groupedWindow[spec] = []*ast.WindowFuncExpr{}
+	}
```

Comment:
How about check whether the array is `nil` in `buildWindowFunctions` rather than set an empty array here?